[Outreachy] Add All Posts Route /notes

RuthNjeri opened the pull request:
Fixes https://github.com/publiclab/plots2/issues/8828

I'm still trying to understand what 'All Posts' means.. this is a first attempt at trying to understand it. 
I'll add a test once feedback is provided.

RuthNjeri commented:
The `dashboard/all` route is meant to display all posts. I have used the activity method to derive the details needed(Research note, questions, wiki, comments tab)

The only pending section is what will be displayed at the banner section...should it be a featured post?

sagarpreet-chadha commented on app/controllers/home_controller.rb:
```diff
@@ -63,6 +63,20 @@ def research
+    if current_user
+      # underscore used for variables not needed
+      # I'm assuming all comments == @comments + @answer_comments
+      # TO DO: Identify what will be pinned at the top of the page. A featured post maybe?
```
I think this TODO is powered by some config maybe as it keeps changes. @jywarren  can confirm

jywarren replied:
I believe this page would not have anything pinned. The dashboard itself would! Thanks!

sagarpreet-chadha commented on app/controllers/home_controller.rb:
```diff
@@ -63,6 +63,20 @@ def research
+      # underscore used for variables not needed
+      # I'm assuming all comments == @comments + @answer_comments
+      # TO DO: Identify what will be pinned at the top of the page. A featured post maybe?
+      _, _, @notes, @wikis, _, @comments, @answer_comments = activity
```
I think this function is being used some other place also and supports pagination also which is great.

jywarren replied:
This is where we only need to show "notes" for the initial view, since unlike the current old dashboard we won't be mixing them together. We'll make separate routes, URLs , and tabs for each type. 

Would this perhaps best fit into... a new controller? I'm not sure. If the home controller isn't too full, i guess it's fine here. Your choice! Thank you!!!

sagarpreet-chadha commented on app/views/dashboard/all_posts.html.erb:
```diff
@@ -0,0 +1 @@
+<h1>All Posts Template</h1>
```
Just a heads up, you can start taking care of internationalization  from beginning :)

jywarren replied:
This will be great not to miss anything. It is possible to do it at the end, but perhaps a better idea to do it as you go, as Sagarpreet suggests!

RuthNjeri replied:
I was planning on creating some FTO issues for translations as well. I'll see how I can implement it partially and create some FTOs if possible.

sagarpreet-chadha reviewed (commented):
Looks great till now

jywarren commented:
Ah, my apologies @RuthNjeri - in this new design, we won't be mixing the types, so each of Notes, Questions, Wikis, and Comments will each simply have their own tab, and their own URL (so we won't load them all on a single page load, it'll be more efficient to just reload the page when the tab is clicked, just like at https://publiclab.org/tag/outreachy for example!). 

Make sense? Thanks!!!

jywarren commented:
So, perhaps we should think of this as a collection of pages that make up "All posts" -- All notes, All questions (would this lead to https://publiclab.org/questions so as not to be redundant?), All wikis (https://publiclab.org/wiki), All comments (this last one might actually link to or replace https://publiclab.org/comments?)

Let's make this PR just about notes, as there is no place where all notes are collected. We can think on the remaining types in the meantime! I'll ping @publiclab/community-reps about this. The main challenge if we link all other types to their corresponding pages, is that if our layout looks like tabs, people may expect those pages to appear like tabs, and be confused if they're led to pages that don't seem to be a sub-tab of the "All posts" page collection. Either this is just OK, and not a big deal, or maybe we would think of some way to indicate that the other types aren't tabs at all, but just links?

RuthNjeri commented:
> Ah, my apologies @RuthNjeri - in this new design, we won't be mixing the types, so each of Notes, Questions, Wikis, and Comments will each simply have their own tab, and their own URL (so we won't load them all on a single page load, it'll be more efficient to just reload the page when the tab is clicked, just like at https://publiclab.org/tag/outreachy for example!).
> 
> Make sense? Thanks!!!

This makes perfect sense!

RuthNjeri commented:
> So, perhaps we should think of this as a collection of pages that make up "All posts" -- All notes, All questions (would this lead to https://publiclab.org/questions so as not to be redundant?), All wikis (https://publiclab.org/wiki), All comments (this last one might actually link to or replace https://publiclab.org/comments?)
> 
> Let's make this PR just about notes, as there is no place where all notes are collected. We can think on the remaining types in the meantime! I'll ping @publiclab/community-reps about this. The main challenge if we link all other types to their corresponding pages, is that if our layout looks like tabs, people may expect those pages to appear like tabs, and be confused if they're led to pages that don't seem to be a sub-tab of the "All posts" page collection. Either this is just OK, and not a big deal, or maybe we would think of some way to indicate that the other types aren't tabs at all, but just links?
> 
> My take is that /questions /wiki /comments are well established and we should try to make the links not appear as tabs, which is a design challenge but not too tough, i think. Let's see what people think!

I'll create a query to collect all notes.. 

About the tabs issue, what if, instead of having a route and page called 'All Posts' we have a dropdown that has a link to all notes, questions, comments, and wiki.. that way, we won't have an entirely new page that looks similar to the topics page.

Since the 'All Posts' link will be on the Dashboard, we can make it a dropdown instead of a link.. unless it is necessary to have a main page for all the posts.

RuthNjeri commented:
@jywarren I have just discovered that there's an action for all notes in the NotesController. The page is https://publiclab.org/notes

This might mean that we have the pages that display all: notes, questions, wikis, and comments. The only pending issue is the design of the page, I have shared my suggestion of using a dropdown on the Dashboard page, see above...

jywarren commented:
Hi @RuthNjeri - 

> @jywarren I have just discovered that there's an action for all notes in the NotesController. The page is https://publiclab.org/notes

Whoa! OK yikes yes! Let's use that, and just refine the design to better match the /tag/____ style. And /notes is a good URL!

> Since the 'All Posts' link will be on the Dashboard, we can make it a dropdown instead of a link..

I guess, to me, it makes sense for the /notes page to easily offer links to the other types, in case you go there and then want to "traverse" to other types. What if we instead of tabs, offered some "pill" links in the same order so it looks familiar in placement and order, but doesn't look like they function as tabs. The row of tabs shown here:

![image](https://user-images.githubusercontent.com/24359/101380795-f65b9380-3883-11eb-9df5-3d0117d85b66.png)

Could instead be a pill nav: https://getbootstrap.com/docs/4.4/components/navs/#pills

![image](https://user-images.githubusercontent.com/24359/101380858-0b382700-3884-11eb-9169-b6ff0359e9dd.png)

That way we can also keep the dashboard simpler with no dropdown. But, we can try this out and re-evaluate once we show it to people -- it would just take some styling changes on the Dashboard to make changes if people really want the dropdown, for example. How does this sound? 

One thing is, I'm not sure if there are functional tests for the /notes route. We may want to add some simple ones. 

Lastly, while we can get rid of all the sidebar content on /notes, we should consider retaining the "by likes" and "by views" options. Maybe those would go in the "sort by" dropdown we had been wanting to get rid of, since they already exist.

Oh-  one more thing -- http://publiclab.org/notes is really slow. I suspect the queries and pagination generating that page are not well optimized. Does anything stand out to you about them? Maybe, we could try to refine them to be closer to the TagController `show` action, but without filtering for tag/topic? I guess it's these three routes:

https://github.com/publiclab/plots2/blob/cbb807ba8e2302f09dafc0060475aa118e34c2c6/app/controllers/notes_controller.rb#L304-L342

RuthNjeri commented:
Using the /notes URL and refining it seems like a great starting point. I think redesigning the page makes a lot of sense.. I'll break down the tasks needed for this in the Planning issue to keep track of the updates and push updates on this PR once I complete the first task...

I'll update you once I need a review....

jywarren commented:
Gosh, actually looking at that code i don't see why it's so inefficient. it seems well limited by timestamp. Ah, i see! `config/routes.rb` shows it's actually the `index` action:

https://github.com/publiclab/plots2/blob/cbb807ba8e2302f09dafc0060475aa118e34c2c6/config/routes.rb#L139

So that's this:

https://github.com/publiclab/plots2/blob/cbb807ba8e2302f09dafc0060475aa118e34c2c6/app/controllers/notes_controller.rb#L6-L9

Leading to this sub-method `set_sidebar` which is really old and probably no longer necessary:

https://github.com/publiclab/plots2/blob/cbb807ba8e2302f09dafc0060475aa118e34c2c6/app/controllers/application_controller.rb#L26-L67

I think we should just skip `set_sidebar` and do it ourselves, since we only need that one query for recent nodes!

jywarren commented:
Ah sorry was still writing into that last comment. But, sounds great, Ruth! Thank you!!!

ebarry commented:
Really great detective work in this thread @RuthNjeri ! Thank you for finding that https://publiclab.org/notes exists, and planning an update!

I just consulted the @publiclab/community-reps staff, and **there is agreement that can we work with pill navs to the pages where we already present content by type.** There was excitement that wiki updates would be easier/more consistent to navigate to. Several people said that it was really the comments that they needed to browse chronologically.

RuthNjeri commented:
Thanks for the clarification @ebarry 😄 

About the need to browse through the comments chronologically, we have the comments route https://publiclab.org/comments where the comments are ordered according to the most recent, is that the chronological order that's needed?

Sagarpreet commented:
> is that the chronological order that's needed

Correct @RuthNjeri 👍

RuthNjeri commented on test/functional/notes_controller_test.rb:
```diff
@@ -700,12 +702,14 @@ def teardown
     expected = [nodes(:one)]
+    # Just that using nodes(:question) is misleading because
+    # the /notes page is designed not to return any question related notes
     questions = [nodes(:question)]
```
same as above

RuthNjeri commented on test/functional/notes_controller_test.rb:
```diff
@@ -690,6 +690,8 @@ def teardown
     expected = [nodes(:one)]
+    # Just that using nodes(:question) is misleading because
+    # the /notes page is designed not to return any question related notes
     questions = [nodes(:question)]
```
# Just that using nodes(:question) is misleading because
    # the /notes page is designed not to return any question related notes

RuthNjeri commented on app/views/notes/index.html.erb:
```diff
@@ -39,13 +39,7 @@
-  <%= render :partial => "sidebar/user" %>
-<% end %>
+<%# The sidebar is also used in the admin/queue section,
+which reuses the 'notes/index' template %>
```
The sidebar is also used in the admin/queue section, which reuses the 'notes/index' template, not sure if it should be removed.

RuthNjeri commented on app/controllers/notes_controller.rb:
```diff
@@ -318,21 +314,15 @@ def liked
-                 .order('created DESC')
-    @unpaginated = true
+    # Not sure why it was unpaginated before
+    @pagy, @notes = pagy(Node.where(type: 'note', status: 1, created: Time.now.to_i - 1.weeks.to_i..Time.now.to_i)
```
Not sure why it was unpaginated before

RuthNjeri commented on app/views/notes/index.html.erb:
```diff
@@ -39,13 +39,5 @@
   <hr />
-  
 </div>
 
```
The sidebar is also used in the admin/queue section, which reuses the 'notes/index' template, not sure if it should be removed.

RuthNjeri replied:
It also removes the write research note button
![Screenshot 2020-12-09 at 22 32 57](https://user-images.githubusercontent.com/7622875/101678068-7f362300-3a6e-11eb-832b-dfc8765e7fe8.png)

RuthNjeri commented on test/functional/notes_controller_test.rb:
```diff
@@ -705,7 +705,7 @@ def teardown
     assert !(notes & questions).present?
```
Just that using nodes(:question) is misleading because
the /notes page is designed not to return any question related notes

RuthNjeri replied:
on line 693 to 695 as well.

RuthNjeri commented:
Hi! This is ready for review. 😄

RuthNjeri commented:
> Hi! This is ready for review. 😄

Tagging @jywarren, @Sagarpreet, and @cesswairimu.. this is ready for review..

cesswairimu reviewed (approved):
Fantastic @RuthNjeri tada: :tada:

Great great job..love the loads of refactoring  and cleanup you did here :rocket: :rocket:
Thanks

cesswairimu commented:
On *popular* and *liked* notes I feel like the website has a lot of data maybe we should increase the limit to 50 or even 100(we are currently showing top 20) and paginate them. What do people think?


Also @RuthNjeri maybe you could make the checklists into issues and link them back to the planning issue maybe smth like what Renisha did here https://github.com/publiclab/plots2/issues/6887#issuecomment-561287047 so that when reviewing its easier to know what part of the project you are working on..also linking the planning issue on PRs will be closed each time we merge and I believe we would like to keep that opened until the project is done. If you have any other methods for doing this please feel free to use them. Thanks and awesome work :tada:

Sagarpreet commented on app/controllers/application_controller.rb:
```diff
@@ -38,13 +38,13 @@ def set_sidebar(type = :generic, data = :all, args = {})
     else # type is generic
       # remove "classroom" postings; also switch to an EXCEPT operator in sql, see https://github.com/publiclab/plots2/issues/375
-      hidden_nids = Node.where(type: :note, status: 1).select { |n| n.has_a_tag('hidden:response') }.collect(&:nid)
+      hidden_nids = Node.hidden_response_nids
```
Nice 👍

RuthNjeri replied:
Thanks!

Sagarpreet commented on app/controllers/application_controller.rb:
```diff
@@ -38,13 +38,13 @@ def set_sidebar(type = :generic, data = :all, args = {})
                else
                  Node.research_notes.joins(:revision).order('node.nid DESC').paginate(page: params[:page])
-      end
+               end
```
Indentation?

RuthNjeri replied:
Yes, to match the if...

Sagarpreet commented on app/controllers/notes_controller.rb:
```diff
@@ -318,24 +313,16 @@ def liked
-      .where(type: 'page', status: 1)
-      .order('nid DESC')
-    @notes = Node.research_notes
+    @notes = published_notes
```
`.limit(20)` will take the first 20 records?
Do we need to sort the data in descending order of Node id (nid)? So we are sure we take latest 20 nodes and then find the popular ones?

RuthNjeri replied:
No limit(20) will not automatically take the first 20 records, I should sort the order then limit..you are right..let me change that..

RuthNjeri replied:
This has been fixed :)

Sagarpreet replied:
Awesome 🎉 
I had faced something similar before, so it immediately came to my notice 😄

Sagarpreet commented on app/controllers/notes_controller.rb:
```diff
@@ -461,4 +448,15 @@ def location
+    hidden_nids = Node.hidden_response_nids
+    notes = Node.research_notes.where('node.status = 1')
+
+    if hidden_nids.empty?
+      notes
+    else
+      notes.where('node.nid NOT IN (?)', hidden_nids)
+    end
```
notes.where('node.nid NOT IN (?)', hidden_nids)
```suggestion
 notes.where('node.nid NOT IN (?)', hidden_nids)
```

RuthNjeri replied:
I noticed that when the hidden_nids are empty, the query does not return the rest of the notes.
The previous query was `notes.where('node.nid NOT IN (?)', hidden_nids) unless hidden_nids`

RuthNjeri replied:
<img width="634" alt="Screenshot 2020-12-10 at 17 19 56" src="https://user-images.githubusercontent.com/7622875/101783948-ee158980-3b0b-11eb-940d-a223bb9f1928.png">

jywarren replied:
aha - were you able to fix that behavior? I think you're right that it's incorrect. Thanks!

RuthNjeri replied:
Currently, the if..else statement works as expected.. if we put this line back 
`notes.where('node.nid NOT IN (?)', hidden_nids) unless hidden_nids` we'll also have to specify what happens if the hidden_nids are empty.. maybe 
```suggestion
   return  notes.where('node.nid NOT IN (?)', hidden_nids) unless hidden_nids
   return  notes if hidden_nids.empty?
```
Which is still similar to the current code...unless we should change it...

Tlazypanda replied:
Hey @RuthNjeri This looks amazing!! :tada: :clap: 

Just noting here can we also check for the performance for line - 
`notes = Node.research_notes.where('node.status = 1')` and then again querying within this depending on `hidden_nids`. If they aren't empty this query may take more time for execution. 

For handling this problem with `notes.where('node.nid NOT IN (?)', hidden_nids)` can we try this once with `where.not` and see if it works? something like `notes.where.not(node.nid: hidden_nids)`

RuthNjeri replied:
Hi @Tlazypanda Thanks so much for the suggestion! It's makes the code so much cleaner!

I've combined the query you've suggested and the notes one into one line
```suggestion
    Node.research_notes
        .where('node.status = 1')
        .where.not(nid: hidden_nids)
```

I'm currently testing it and it seems that it's working.. pushing it in a few... it might also solve the performance issue of running two queries separately?

Sagarpreet commented on app/models/node.rb:
```diff
@@ -124,7 +124,11 @@ def slug_from_path
+    tags.where(name: name).size.positive?
+  end
+
+  def self.hidden_response_nids
```
Maybe use `hidden_response_node_ids `?

RuthNjeri replied:
That makes sense, I'll modify it.

RuthNjeri replied:
Done!

Sagarpreet commented on app/models/node.rb:
```diff
@@ -920,8 +924,14 @@ def self.research_notes
+    notes = Node.where(type: 'note')
+
+    # if the nids are empty, it should not be used in the query
+    if nids.empty?
+      notes
+    else
+      notes.where('node.nid NOT IN (?)', nids)
+    end
```
`notes.where('node.nid NOT IN (?)', nids)` ?

RuthNjeri replied:
thanks, see comment [here](https://github.com/publiclab/plots2/pull/8800/files/c5954b67e9b5c1e72d6e6192a73883f167362c87#r540164441) on the same

Sagarpreet commented on app/models/node.rb:
```diff
@@ -920,8 +924,14 @@ def self.research_notes
-                .where('node.nid NOT IN (?)', nids)
+    notes = Node.where(type: 'note')
+
+    # if the nids are empty, it should not be used in the query
```
Can you also write why we donot have to use `nids` in comment?

RuthNjeri replied:
sure!

RuthNjeri replied:
Done!

Sagarpreet commented on app/views/wiki/_wikis.html.erb:
```diff
@@ -29,7 +29,7 @@
 
 <% if @pagy %>
-  <%== pagy_bootstrap_nav @pagy %>
+  <%= pagy_bootstrap_nav @pagy %>
```
Good catch 🎉

Sagarpreet reviewed (commented):
Great work 💯

cesswairimu commented on app/controllers/notes_controller.rb:
```diff
@@ -318,24 +313,16 @@ def liked
-      .order('nid DESC')
-    @notes = Node.research_notes
+    @notes = published_notes
       .limit(20)
```
seems like here the `.order` should come before the `limit`

RuthNjeri replied:
Yes, good catch, I'll add that

RuthNjeri replied:
I've just realized that the order method is from the [@wiki ](https://github.com/publiclab/plots2/pull/8800/files/c5954b67e9b5c1e72d6e6192a73883f167362c87#diff-c254740097a8c0601832077488db77e572b2c2b926f77508fadc6081b0052804L333)query, but it's true that the order should come before the limit in order to get the most recent ones which is also what @Sagarpreet has suggested.

RuthNjeri replied:
I've added the order in the [published_notes](https://github.com/publiclab/plots2/pull/8800/files#diff-c254740097a8c0601832077488db77e572b2c2b926f77508fadc6081b0052804R454) method

RuthNjeri commented:
> On _popular_ and _liked_ notes I feel like the website has a lot of data maybe we should increase the limit to 50 or even 100(we are currently showing top 20) and paginate them. What do folks think about this? @ebarry @jywarren @sagarpreet-chadha
> 
> Also @RuthNjeri maybe you could make the checklists into issues and link them back to the planning issue maybe smth like what Renisha did here [#6887 (comment)](https://github.com/publiclab/plots2/issues/6887#issuecomment-561287047) so that when reviewing its easier to know what part of the project you are working on..also linking the planning issue on PRs will be closed each time we merge and I believe we would like to keep that opened until the project is done. If you have any other methods for doing this please feel free to use them. Thanks and awesome work 🎉

Thanks @cesswairimu , I'll break down the checklists into issues, it makes much more sense.
I'll wait for @ebarry @jywarren @sagarpreet-chadha to chip in on the limit suggestion as I work on the rest of the feedback.

RuthNjeri commented:
@Sagarpreet , @cesswairimu about the limit and order methods for the popular which also affects recent and the liked actions..

I've noticed that when I first order all the queries in descending order and then order the cached_liked in DESC as well, the notes are displayed according to the most recent only.

If an older note has the highest like, it will still come after the most recent one..
![Screenshot 2020-12-10 at 18 10 38](https://user-images.githubusercontent.com/7622875/101790008-03da7d00-3b13-11eb-9be0-e38b19525b9c.png)
<img width="1309" alt="Screenshot 2020-12-10 at 18 10 23" src="https://user-images.githubusercontent.com/7622875/101790024-08069a80-3b13-11eb-8bc4-9a4f8f6929e1.png">

RuthNjeri commented on test/functional/notes_controller_test.rb:
```diff
@@ -692,7 +692,7 @@ def teardown
     questions = [nodes(:question)]
     assert (notes & expected).present?
-    assert (notes & questions).present?
+    assert !(notes & questions).present?
```
This was a failing test before...

cesswairimu commented on app/controllers/notes_controller.rb:
```diff
@@ -461,4 +448,15 @@ def location
+
+  def published_notes
+    hidden_nids = Node.hidden_response_node_ids
+    notes = Node.research_notes.where('node.status = 1').order('node.nid DESC')
```
@RuthNjeri I think it could be its because you are already doing an order by nid on `published_notes` on line 454, maybe we could remove the order here and just do it on the methods...what do you think?

RuthNjeri replied:
Hey @cesswairimu in this [version](https://github.com/publiclab/plots2/pull/8800#issuecomment-742583719) (the screenshot), I did that, I only pushed the current one (line 454) after I tested it.. the results are the same.. 

I've attached other screenshots..

<img width="784" alt="Screenshot 2020-12-10 at 20 41 46" src="https://user-images.githubusercontent.com/7622875/101809151-40fd3a00-3b28-11eb-82f1-abb96a53d050.png">

<img width="597" alt="Screenshot 2020-12-10 at 20 42 05" src="https://user-images.githubusercontent.com/7622875/101809161-43f82a80-3b28-11eb-8db1-321fb2c2bbb6.png">

![Screenshot 2020-12-10 at 20 42 13](https://user-images.githubusercontent.com/7622875/101809218-54100a00-3b28-11eb-8b14-cefa4ab2319b.png)

RuthNjeri replied:
Maybe the node.nid order takes precedence?

cesswairimu replied:
I believe what the prob is the double `.order`...we could try (with your `published_notes` method as the above screenshot )
```
def liked
    @title = I18n.t('notes_controller.highly_liked_research_notes')
    @notes = published_notes
             .order('cached_likes DESC').limit 20
    @unpaginated = true
    render template: 'notes/index'
  end
```
I don't think we need to order by latest on most liked...then you can have the order on `recent` what do you think?

cesswairimu replied:
Also just seen this for multiple ordering...maybe you could try it out `.order(cached_likes: :desc, nid: :desc)`

RuthNjeri replied:
Thanks @cesswairimu let me run it in the debugger and share the results

RuthNjeri replied:
For this  `I don't think we need to order by latest on most liked...then you can have the order by nid on recent what do you think?` It makes sense, though with the limit of 20 nodes, I think it defaults in picking the earliest 20 nodes...

RuthNjeri replied:
@cesswairimu the query you shared works as expected 😄 , I think the issue was with the way the arguments within the order params were positioned.. in the first query it was
`Node.research_notes.where('node.status = 1').order('node.nid DESC').order('cached_likes DESC')`
The node.nid order is positioned first so the default order will be according to the` node.nid DESC`

And the second one that you've suggested `Node.research_notes.where('node.status = 1').order(cached_likes: :desc, nid: :desc)`
The node.nid order is positioned second.. and it prioritizes the cached_likes

<img width="925" alt="Screenshot 2020-12-10 at 22 09 27" src="https://user-images.githubusercontent.com/7622875/101818060-6132f600-3b34-11eb-8966-80b9d81f03b7.png">

(the order is based on the most recent with the highest number of likes)


I've tested it with the same query you shared but switched up the position and got the same original problem,
`Node.research_notes.where('node.status = 1').order(nid: :desc, cached_likes: :desc, )`

<img width="888" alt="Screenshot 2020-12-10 at 22 11 26" src="https://user-images.githubusercontent.com/7622875/101818268-abb47280-3b34-11eb-9ee1-c210cf0e5189.png">

(the order is based on the most recent and not exactly on the highest number of likes)... still confusing to me but so far this is the only thing that makes sense..


It's now fixed 🎉 , Thanks for the help @cesswairimu 😄

jywarren commented:
Just chiming in to say that this is looking great, and I appreciate all the collab and discussion and thought going on. Some refinements like pagination could be done in a follow-up PR if you prefer, but it's your choice! In general we love to do things in discrete smaller PRs but if you're deep into it already, no worries! Thanks, everyone!

Also, I noted some issues with running the tests in GitPod in [this comment](https://github.com/publiclab/plots2/pull/8801#issuecomment-742730049) so i wonder if the same is true here-- but until our Travis tests are fixed in https://github.com/publiclab/plots2/pull/8795 we may need to rely on manually running tests in the standard GitPod environment as a substitute, so if it's possible to confirm that they have passed as a final step before we mark this (and other) PRs for review, that would be great until we finally complete https://github.com/publiclab/plots2/pull/8795. Thank you for your patience with this!!!

RuthNjeri commented:
Hi, @jywarren 😄  I'm okay, waiting for the tests to work, I've fixed the order issue and added the limit number.. 
In the meantime, I can begin working on the design from another branch..

RuthNjeri commented:
I've confirmed that all tests running in the `notes_controller_test` are all passing.

RuthNjeri commented:
@jywarren this PR has all tests ready to be merged. The `node_test.rb` and `notes_controller_test.rb` files are all passing.

Please have a look at the hidden_nids query and a test I've just updated [here](https://github.com/publiclab/plots2/pull/8800/files#diff-6997c2ce770b648ac2c34043b6b33b6f659fc3aa24fa82fc6afd7f4ed05c7805R511) to know if it's good to merge or needs a few more revisions... thanks!

jywarren commented on app/models/node.rb:
```diff
@@ -123,8 +123,13 @@ def slug_from_path
+  def self.hidden_response_node_ids
+    Node.joins(:node_tag)
+        .joins('LEFT OUTER JOIN term_data ON term_data.tid = community_tags.tid')
+        .select('node.nid')
```
Looks great! However i think this may need more selected fields -- https://guides.rubyonrails.org/active_record_querying.html#selecting-specific-fields says it actually changes the `SELECT X` section of the SQL query, so we may need all those used in the joins and where fields.

I read your response about this somewhere, i think... but can't find it now? Maybe you deleted it, or was it in Gitter? gosh, my brain isn't working so clearly right now as I'm deep in this Travis thing... sorry!

Just want to confirm this here though after checking the Rails docs!

RuthNjeri replied:
It was in my other [PR](https://github.com/publiclab/plots2/pull/8848#discussion_r543489655) 😄 

Thanks, I'll have a look at the doc...and update the PR.

jywarren commented:
Ah, and regarding the CodeClimate recommendations, if you click on the CodeClimate check it'll make suggestions, and if you want to ignore them (which is ok! Your judgement!) i can "skip" them, so just tell me if that's what you prefer! Thanks!

RuthNjeri commented:
> Ah, and regarding the CodeClimate recommendations, if you click on the CodeClimate check it'll make suggestions, and if you want to ignore them (which is ok! Your judgement!) i can "skip" them, so just tell me if that's what you prefer! Thanks!
> 
> (actually i went ahead and bypassed them just to make things simpler for the next step, but if you want you can still address them and i think they're still listed!)

It's okay to ignore the CodeClimate issues, I don't think I can solve them...

RuthNjeri commented:
The current test failing here is also failing in the Main branch.. https://github.com/publiclab/plots2/runs/1561451524

RuthNjeri commented:
The force-push is due to the recent rebase...

jywarren commented:
OK, this should be good now if rebased. Thank you for your patience!!!

RuthNjeri commented:
> OK, this should be good now if rebased. Thank you for your patience!!!

Thanks @jywarren, we should actually merge this[ PR](https://github.com/publiclab/plots2/pull/8800) first, it has all the up to date changes and all the feedback implemented.

This current PR still has some pending feedback I need to implement.

RuthNjeri commented:
Force pushed due to recent rebase with the main branch

RuthNjeri commented:
Hi @jywarren, it seems that the database setup is failing when the build is running...
<img width="521" alt="Screenshot 2020-12-16 at 22 04 51" src="https://user-images.githubusercontent.com/7622875/102394454-d302b800-3fea-11eb-8e49-06001f548ae3.png">

I'm not sure why it's failing...

jywarren commented:
Strange, ok I'm actually seeing this integration test fail:

```
FAIL["test_should_choose_i18n_for_sidebar/_author_+_sidebar/_post_button", #<Minitest::Reporters::Suite:0x000055bd75e04178 @name="I18nTest">, 129.59562038299998]
 test_should_choose_i18n_for_sidebar/_author_+_sidebar/_post_button#I18nTest (129.60s)
        Expected at least 1 element matching "h4", found 0..
        Expected 0 to be >= 1.
        test/integration/I18n_test.rb:379:in `block (2 levels) in <class:I18nTest>'
        test/integration/I18n_test.rb:365:in `each'
        test/integration/I18n_test.rb:365:in `block in <class:I18nTest>'

=============|
```

I restarted to see if it's intermittent... Any ideas otherwise?

cesswairimu commented on app/views/notes/index.html.erb:
```diff
@@ -39,13 +39,5 @@
-  
 </div>
 
-<% if params[:controller] == "tag" || params[:controller] == "search" %>
-  <%= render :partial => "sidebar/related" %>
-<% elsif params[:action] == "author" || params[:action] == "author_topic" %>
-  <%= render :partial => "sidebar/author" %>
-<% else %>
-  <%= render :partial => "sidebar/user" %>
-<% end %>
```
Hey @RuthNjeri the removal of this section is the one causing the tests to fail. Also I  don't think  we want to remove this sidebar...it gives access to recent wiki edits ...You might also want to add back the `set_sidebar` under  index in the controller. what do you think?

RuthNjeri replied:
Thanks @cesswairimu 😄 

The removal of that section was part of this [discussion](https://github.com/publiclab/plots2/pull/8800#issuecomment-740052184)  with @jywarren where we replace it with the sort dropdown for the notes...

I think a user will still be able to access the recent wiki edits via the wiki section that will be linked on the notes page, I am working on this [here](https://github.com/publiclab/plots2/pull/8848), therefore, access to the recent edits won't be completely lost, just changed to another page....

I think this can also be open to further discussion before the PR is merged..

cesswairimu replied:
Oh okay, then I think we might need to rewrite that test or remove it

RuthNjeri commented:
> Strange, ok I'm actually seeing this integration test fail:
> 
> ```
> FAIL["test_should_choose_i18n_for_sidebar/_author_+_sidebar/_post_button", #<Minitest::Reporters::Suite:0x000055bd75e04178 @name="I18nTest">, 129.59562038299998]
>  test_should_choose_i18n_for_sidebar/_author_+_sidebar/_post_button#I18nTest (129.60s)
>         Expected at least 1 element matching "h4", found 0..
>         Expected 0 to be >= 1.
>         test/integration/I18n_test.rb:379:in `block (2 levels) in <class:I18nTest>'
>         test/integration/I18n_test.rb:365:in `each'
>         test/integration/I18n_test.rb:365:in `block in <class:I18nTest>'
> 
> =============|
> ```
> 
> I restarted to see if it's intermittent... Any ideas otherwise?

Hi @jywarren the test is related to the sidebar that I removed, so I've removed the test...though there's a discussion happening above whether the sidebar should be removed..

jywarren commented:
No I think that's correct to remove the sidebar and the test, thank you! This looks awesome and it's so comforting to see the green All checks have passed ✅ message 😅

Merging now! Congratulations!!!! Thanks everyone!!!

jywarren commented:
This should shortly be visible at stable.publiclab.org!!!

RuthNjeri commented:
Nice! The notes URL is loading a bit faster now 😄

jywarren commented:
Awesome to see!

On Thu, Dec 17, 2020, 1:52 PM Ruth <notifications@github.com> wrote:

> Nice! The notes URL is loading a bit faster now 😄
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/8800#issuecomment-747630565>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J7ITCUIJMR76VG6EL3SVJHOVANCNFSM4UNRMNYQ>
> .
>
